diff --git a_bak/fs/namespace.c b/fs/namespace.c
index edb8581..67c852c 100644
--- a_bak/fs/namespace.c
+++ b/fs/namespace.c
@@ -1118,6 +1118,11 @@ struct vfsmount *vfs_kern_mount(struct file_system_type *type,
 {
 	struct fs_context *fc;
 	struct vfsmount *mnt;
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+	struct mount *m;
+	struct mnt_namespace *mnt_ns;
+	int mnt_id;
+#endif
 	int ret = 0;
 
 	if (!type)
diff --git a_bak/include/linux/sched.h b/include/linux/sched.h
index c28533c..74fe1d4 100644
--- a_bak/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1469,16 +1469,8 @@ struct task_struct {
 	ANDROID_KABI_RESERVE(4);
 	ANDROID_KABI_RESERVE(5);
 	ANDROID_KABI_RESERVE(6);
-#ifdef CONFIG_KSU_SUSFS
-	ANDROID_KABI_USE(7, u64 susfs_task_state);
-#else
 	ANDROID_KABI_RESERVE(7);
-#endif
-#ifdef CONFIG_KSU_SUSFS
-	ANDROID_KABI_USE(8, u64 susfs_last_fake_mnt_id);
-#else
 	ANDROID_KABI_RESERVE(8);
-#endif
 
 	/*
 	 * New fields for task_struct should be added above here, so that
