--- a/OplusKernelEnvConfig.mk	2026-03-18 22:35:04.319411564 +0800
+++ b/OplusKernelEnvConfig.mk	2026-03-18 22:35:35.720950945 +0800
@@ -15,6 +15,8 @@
 ### ------------------------------------------------------------------------------
 ##################################################################################
 
+export BRAND_SHOW_FLAG=oneplus
+
 -include oplus_native_features.mk
 
 ###ifdef OPLUS_ARCH_INJECT
