diff --git "a/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\xiaomi_fix_work\\orig\\kernel\\kallsyms.c" "b/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\xiaomi_fix_work\\new\\kernel\\kallsyms.c"
index ec185f2..7a0f870 100644
--- a/kernel/kallsyms.c
+++ b/kernel/kallsyms.c
@@ -23,6 +23,9 @@
 #include <linux/slab.h>
 #include <linux/filter.h>
 #include <linux/compiler.h>
+#ifdef CONFIG_KSU_SUSFS_HIDE_KSU_SUSFS_SYMBOLS
+#include <linux/susfs_def.h>
+#endif // #ifdef CONFIG_KSU_SUSFS_HIDE_KSU_SUSFS_SYMBOLS
 
 /*
  * These will be re-linked against their real values
diff --git "a/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\xiaomi_fix_work\\orig\\mm\\memory.c" "b/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\xiaomi_fix_work\\new\\mm\\memory.c"
index a7f5a38..d4164e7 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -79,6 +79,10 @@
 #include <asm/tlbflush.h>
 #include <asm/pgtable.h>
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+#include <linux/susfs_def.h>
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MAP
+
 #include "internal.h"
 
 #if defined(LAST_CPUPID_NOT_IN_PAGE_FLAGS) && !defined(CONFIG_COMPILE_TEST)
