diff --git "a/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\_fx\\orig\\fs\\namespace.c" "b/C:\\Users\\29907\\Desktop\\Ai\\GKI\\_tools\\research\\_fx\\new\\fs\\namespace.c"
index 75f1edc..e28a7be 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -30,6 +30,9 @@
 #include <uapi/linux/mount.h>
 #include <linux/fs_context.h>
 #include <linux/shmem_fs.h>
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+#include <linux/susfs_def.h>
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 
 #include "pnode.h"
 #include "internal.h"
@@ -38,6 +41,18 @@
 
 
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+extern bool susfs_is_current_ksu_domain(void);
+extern struct static_key_true susfs_is_sdcard_android_data_not_decrypted;
+
+#define CL_COPY_MNT_NS BIT(25) /* used by copy_mnt_ns() */
+
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+
+
+
+
+
 /* Maximum number of mounts in a mount namespace */
 unsigned int sysctl_mount_max __read_mostly = 100000;
 
