--- a/KernelSU/kernel/kernel_compat.h	2025-10-01 15:27:53.385016451 +0800
+++ b/KernelSU/kernel/kernel_compat.h	2025-10-01 15:44:09.271953898 +0800
@@ -73,10 +73,4 @@
 extern ssize_t ksu_kernel_write_compat(struct file *p, const void *buf,
 				       size_t count, loff_t *pos);
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 0, 0)
-#define ksu_access_ok(addr, size)	access_ok(addr, size)
-#else
-#define ksu_access_ok(addr, size)	access_ok(VERIFY_READ, addr, size)
-#endif
-
 #endif
