diff --git a/include/linux/sched.h b/include/linux/sched.h
index 550d245..7732b85 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1262,9 +1262,6 @@ struct task_struct {
 #endif
 #ifdef CONFIG_PERF_EVENTS
 	struct perf_event_context	*perf_event_ctxp[perf_nr_task_contexts];
-#if defined(CONFIG_KSU_SUSFS)
-	u64 susfs_task_state;
-#endif
 	struct mutex			perf_event_mutex;
 	struct list_head		perf_event_list;
 #endif
