--- a/include/linux/sched.h	2025-09-18 16:07:01.000000000 +0800
+++ b/include/linux/sched.h	2025-09-18 16:11:57.604827919 +0800
@@ -1470,11 +1470,7 @@
 	ANDROID_KABI_RESERVE(5);
 	ANDROID_KABI_RESERVE(6);
 	ANDROID_KABI_RESERVE(7);
-#ifdef CONFIG_KSU_SUSFS
-	ANDROID_KABI_USE(8, u64 susfs_last_fake_mnt_id);
-#else
 	ANDROID_KABI_RESERVE(8);
-#endif
 
 	/*
 	 * New fields for task_struct should be added above here, so that
