diff --git a/fs/namespace.c b/fs/namespace.c
index c3005b5..832e7c9 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1073,7 +1073,7 @@ bypass_orig_flow:
 	mnt->mnt_parent		= mnt;
 
 #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
-	// - If caller process is zygote, then it is a normal mount, so we calculate the next available 
+	// - If caller process is zygote, then it is a normal mount, so we calculate the next available
 	//   fake mnt_id for this mount
 	if (susfs_is_current_zygote_domain()) {
 		mnt_ns = current->nsproxy->mnt_ns;
@@ -1118,6 +1118,11 @@ struct vfsmount *vfs_kern_mount(struct file_system_type *type,
 {
 	struct fs_context *fc;
 	struct vfsmount *mnt;
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+	struct mount *m;
+	struct mnt_namespace *mnt_ns;
+	int mnt_id;
+#endif
 	int ret = 0;
 
 	if (!type)
@@ -1172,7 +1177,7 @@ static struct mount *clone_mnt(struct mount *old, struct dentry *root,
 	bool is_current_ksu_domain = susfs_is_current_ksu_domain();
 	bool is_current_zygote_domain = susfs_is_current_zygote_domain();
 
-	/* - It is very important that we need to use CL_COPY_MNT_NS to identify whether 
+	/* - It is very important that we need to use CL_COPY_MNT_NS to identify whether
 	 *   the clone is a copy_tree() or single mount like called by __do_loopback()
 	 * - if caller process is KSU, consider the following situation:
 	 *     1. it is NOT doing unshare => call alloc_vfsmnt() to assign a new sus mnt_id
@@ -1978,7 +1983,7 @@ struct mount *copy_tree(struct mount *mnt, struct dentry *dentry,
 #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 	bool is_current_zygote_domain = susfs_is_current_zygote_domain();
 #endif
-	
+
 	if (!(flag & CL_COPY_UNBINDABLE) && IS_MNT_UNBINDABLE(mnt))
 		return ERR_PTR(-EINVAL);
 
@@ -2033,7 +2038,7 @@ struct mount *copy_tree(struct mount *mnt, struct dentry *dentry,
 					attach_mnt_count++;
 					q->mnt_id += attach_mnt_count;
 			}
-#endif				
+#endif
 			lock_mount_hash();
 			list_add_tail(&q->mnt_list, &res->mnt_list);
 			attach_mnt(q, parent, p->mnt_mp);
@@ -3586,7 +3591,7 @@ struct mnt_namespace *copy_mnt_ns(unsigned long flags, struct mnt_namespace *ns,
 	// q->mnt.susfs_mnt_id_backup -> original mnt_id
 	// q->mnt_id -> will be modified to the fake mnt_id
 
-	// Here We are only interested in processes of which original mnt namespace belongs to zygote 
+	// Here We are only interested in processes of which original mnt namespace belongs to zygote
 	// Also we just make use of existing 'q' mount pointer, no need to delcare extra mount pointer
 	if (is_zygote_pid) {
 		last_entry_mnt_id = list_first_entry(&new_ns->list, struct mount, mnt_list)->mnt_id;
diff --git a/include/linux/sched.h b/include/linux/sched.h
index a3c8289..b80066a 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1474,11 +1474,7 @@ struct task_struct {
 #else
 	ANDROID_KABI_RESERVE(7);
 #endif
-#ifdef CONFIG_KSU_SUSFS
-	ANDROID_KABI_USE(8, u64 susfs_last_fake_mnt_id);
-#else
 	ANDROID_KABI_RESERVE(8);
-#endif
 
 	/*
 	 * New fields for task_struct should be added above here, so that
