--- a/fs/proc/cmdline.c	2025-06-15 18:49:30.857056686 +0800
+++ b/fs/proc/cmdline.c	2025-06-15 18:49:59.424856774 +0800
@@ -10,8 +10,18 @@
 static char patched_cmdline[COMMAND_LINE_SIZE];
 #endif
 
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+extern int susfs_spoof_cmdline_or_bootconfig(struct seq_file *m);
+#endif
+
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	if (!susfs_spoof_cmdline_or_bootconfig(m)) {
+		seq_putc(m, '\n');
+		return 0;
+	}
+#endif
 #ifndef CONFIG_PROC_BEGONIA_CMDLINE
 	seq_printf(m, "%s\n", saved_command_line);
 #else
