diff --git a/k20p_bak/fs/proc/cmdline.c b/k20p/fs/proc/cmdline.c
index 4a24f49..b715599 100644
--- a/k20p_bak/fs/proc/cmdline.c
+++ b/k20p/fs/proc/cmdline.c
@@ -27,8 +27,18 @@ static void proc_command_line_init(void) {
 }
 #endif
 
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+extern int susfs_spoof_cmdline_or_bootconfig(struct seq_file *m);
+#endif
+
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	if (!susfs_spoof_cmdline_or_bootconfig(m)) {
+		seq_putc(m, '\n');
+		return 0;
+	}
+#endif
 #ifdef CONFIG_INITRAMFS_IGNORE_SKIP_FLAG
 	seq_printf(m, "%s\n", proc_command_line);
 #else
diff --git a/k20p_bak/fs/proc/task_mmu.c b/k20p/fs/proc/task_mmu.c
index 9f2db87..d83f21e 100644
--- a/k20p_bak/fs/proc/task_mmu.c
+++ b/k20p/fs/proc/task_mmu.c
@@ -21,6 +21,9 @@
 #include <linux/mm_inline.h>
 #include <linux/ctype.h>
 #include <linux/mm_inline.h>
+#if defined(CONFIG_KSU_SUSFS_SUS_KSTAT) || defined(CONFIG_KSU_SUSFS_SUS_MAP)
+#include <linux/susfs_def.h>
+#endif
 
 #include <asm/elf.h>
 #include <asm/tlb.h>
diff --git a/k20p_bak/kernel/sys.c b/k20p/kernel/sys.c
index 46abe65..1b15239 100644
--- a/k20p_bak/kernel/sys.c
+++ b/k20p/kernel/sys.c
@@ -1212,6 +1212,9 @@ SYSCALL_DEFINE1(newuname, struct new_utsname __user *, name)
 
 	down_read(&uts_sem);
 	memcpy(&tmp, utsname(), sizeof(tmp));
+#ifdef CONFIG_KSU_SUSFS_SPOOF_UNAME
+	susfs_spoof_uname(&tmp);
+#endif
 	up_read(&uts_sem);
 
 	rcu_read_lock();
