diff --git a/fs/notify/fdinfo.c b/fs/notify/fdinfo.c
index 0853f96..4c0d5cb 100644
--- a/fs/notify/fdinfo.c
+++ b/fs/notify/fdinfo.c
@@ -107,6 +107,7 @@ static void inotify_fdinfo(struct seq_file *m, struct fsnotify_mark *mark)
 		 * least one bit (FS_EVENT_ON_CHILD) which is
 		 * used only internally to the kernel.
 		 */
+		u32 mask = mark->mask & IN_ALL_EVENTS;
 #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 		if (likely(current->susfs_task_state & TASK_STRUCT_NON_ROOT_USER_APP_PROC) &&
 				unlikely(inode->i_state & INODE_STATE_SUS_KSTAT)) {
@@ -137,7 +138,6 @@ out_free_pathname:
 		}
 out_seq_printf:
 #endif
-		u32 mask = mark->mask & IN_ALL_EVENTS;
 		seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:%x ",
 			   inode_mark->wd, inode->i_ino, inode->i_sb->s_dev,
 			   mask, mark->ignored_mask);
diff --git a/fs/readdir.c b/fs/readdir.c
index e016116..130d101 100644
