diff --git a/fs/namespace.c b/fs/namespace.c
index e3d593c..063a83d 100644
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1116,6 +1116,12 @@ struct vfsmount *vfs_kern_mount(struct file_system_type *type,
 	struct vfsmount *mnt;
 	int ret = 0;
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+	struct mount *m;
+	struct mnt_namespace *mnt_ns;
+	int mnt_id;
+#endif
+
 	if (!type)
 		return ERR_PTR(-EINVAL);
 
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 9d583db..0a55b7d 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1320,11 +1320,7 @@ struct task_struct {
 #else
 	ANDROID_KABI_RESERVE(7);
 #endif
-#ifdef CONFIG_KSU_SUSFS
-	ANDROID_KABI_USE(8, u64 susfs_last_fake_mnt_id);
-#else
 	ANDROID_KABI_RESERVE(8);
-#endif
 
 	/*
 	 * New fields for task_struct should be added above here, so that
