--- a/drivers/android/binder_alloc.c	2025-07-01 17:02:45.125407997 +0800
+++ b/drivers/android/binder_alloc.c	2025-07-01 17:04:12.032559970 +0800
@@ -472,6 +472,21 @@
 		return ERR_PTR(-EINVAL);
 	}
 
+	if (is_async
+		&& (alloc->free_async_space < 3 * (size + sizeof(struct binder_buffer))
+		|| (alloc->free_async_space < REKERNEL_WARN_AHEAD_SPACE))) {
+		rcu_read_lock();
+		proc_task = find_task_by_vpid(alloc->pid);
+		rcu_read_unlock();
+		if (proc_task != NULL && start_rekernel_server() == 0) {
+			if (line_is_frozen(proc_task)) {
+     			char binder_kmsg[REKERNEL_PACKET_SIZE];
+                snprintf(binder_kmsg, sizeof(binder_kmsg), "type=Binder,bindertype=free_buffer_full,oneway=1,from_pid=%d,from=%d,target_pid=%d,target=%d;", current->pid, task_uid(current).val, proc_task->pid, task_uid(proc_task).val);
+         		send_netlink_message(binder_kmsg, strlen(binder_kmsg));
+			}
+		}
+	}
+
 	/* Pad 0-size buffers so they get assigned unique addresses */
 	size = max(size, sizeof(void *));
 
