--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -927,7 +927,7 @@
 #ifdef CONFIG_KSU_SUSFS_SUS_MAP
 	if (vma->vm_file) {
 		if (SUSFS_IS_INODE_SUS_MAP(file_inode(vma->vm_file)))
-			return;
+			return 0;
 	}
 #endif // #ifdef CONFIG_KSU_SUSFS_SUS_MAP
 

--- a/fs/notify/fdinfo.c
+++ b/fs/notify/fdinfo.c
@@ -105,6 +105,31 @@
 	inode_mark = container_of(mark, struct inotify_inode_mark, fsn_mark);
 	inode = igrab(fsnotify_conn_inode(mark->connector));
 	if (inode) {
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+		mnt = real_mount(file->f_path.mnt);
+		if (mnt->mnt_id >= DEFAULT_KSU_MNT_ID &&
+			likely(susfs_is_current_proc_umounted()))
+		{
+			struct path path;
+			char *pathname = kmalloc(PAGE_SIZE, GFP_KERNEL);
+			char *dpath;
+
+			dpath = d_path(&file->f_path, pathname, PAGE_SIZE);
+			if (pathname && !IS_ERR(dpath) && !kern_path(dpath, 0, &path) &&
+			    path.dentry->d_inode) {
+				seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:0 ",
+						inode_mark->wd, path.dentry->d_inode->i_ino,
+						path.dentry->d_inode->i_sb->s_dev,
+						mark->mask & IN_ALL_EVENTS);
+				show_mark_fhandle(m, path.dentry->d_inode);
+				seq_putc(m, '\n');
+				path_put(&path);
+			}
+			kfree(pathname);
+			iput(inode);
+			return;
+		}
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 		/*
 		 * IN_ALL_EVENTS represents all of the mask bits
 		 * that we expose to userspace.  There is at
@@ -130,7 +155,11 @@
 
 #ifdef CONFIG_FANOTIFY
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+static void fanotify_fdinfo(struct seq_file *m, struct fsnotify_mark *mark, struct file *file)
+#else
 static void fanotify_fdinfo(struct seq_file *m, struct fsnotify_mark *mark)
+#endif
 {
 	unsigned int mflags = 0;
 	struct inode *inode;

--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -30,10 +30,21 @@
 #include <uapi/linux/mount.h>
 #include <linux/fs_context.h>
 #include <linux/shmem_fs.h>
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+#include <linux/susfs_def.h>
+#endif
 
 #include "pnode.h"
 #include "internal.h"
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+extern bool susfs_is_current_ksu_domain(void);
+extern struct static_key_true susfs_is_sdcard_android_data_not_decrypted;
+
+#define CL_COPY_MNT_NS BIT(25) /* used by copy_mnt_ns() */
+
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+
 
 
 

--- a/include/trace/hooks/sched.h
+++ b/include/trace/hooks/sched.h
@@ -95,7 +95,7 @@
 #define trace_android_rvh_prepare_prio_fork(p)
 #define trace_android_rvh_finish_prio_fork(p)
 #define trace_android_rvh_rtmutex_prepare_setprio(p, pi_task)
-#define trace_android_rvh_set_user_nice(p, nice)
+#define trace_android_rvh_set_user_nice(p, nice, allowed)
 #define trace_android_rvh_setscheduler(p)
 #define trace_android_rvh_check_preempt_wakeup(p, ignore)
 #define trace_android_rvh_check_preempt_tick(p, ideal_runtime)

--- a/include/trace/hooks/vh_af_packet.h
+++ b/include/trace/hooks/vh_af_packet.h
@@ -25,7 +25,7 @@
 		int *do_drop),
 	TP_ARGS(sk, skb, dev, pt, do_drop));
 #else
-#define android_vh_check_dhcp_pkt(sk, skb, dev, pt, do_drop)
+#define trace_android_vh_check_dhcp_pkt(sk, skb, dev, pt, do_drop)
 #endif
 #endif /* _TRACE_HOOK_AF_PACKET_H */
 /* This part must be outside protection */

--- a/kernel/taskstats.c
+++ b/kernel/taskstats.c
@@ -466,7 +466,7 @@
 	}
 	stats->zram_compressed = K(zspages);
 }
-#elif
+#else
 static void sysstats_fill_zoneinfo(struct sys_memstats *stats)
 {
 }

--- a/arch/arm64/configs/defconfig
+++ b/arch/arm64/configs/defconfig
@@ -703,7 +703,7 @@
 CONFIG_TI_SCI_CLK=y
 CONFIG_COMMON_CLK_QCOM=y
 CONFIG_QCOM_A53PLL=y
-CONFIG_QCOM_CLK_APCS_MSM8916=y
+# CONFIG_QCOM_CLK_APCS_MSM8916 is not set
 CONFIG_QCOM_CLK_SMD_RPM=y
 CONFIG_QCOM_CLK_RPMH=y
 CONFIG_IPQ_GCC_8074=y
