--- a/drivers/android/binderfs.c	2025-09-17 00:07:52.318423989 +0800
+++ b/drivers/android/binderfs.c	2025-09-17 00:06:35.131687650 +0800
@@ -1,6 +1,6 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 
-#include <linux/compiler_type.h>
+#include <linux/compiler.h>
 #include <linux/errno.h>
 #include <linux/fs.h>
 #include <linux/fsnotify.h>
@@ -35,6 +35,9 @@
 
 #include "binder_internal.h"
 
+#define ida_alloc_max(a, b, c) ida_simple_get(a, 0, b + 1, c)
+#define ida_free ida_remove
+
 #define FIRST_INODE 1
 #define SECOND_INODE 2
 #define INODE_OFFSET 3
@@ -675,7 +678,7 @@
 	 * allowed to do. So removing the SB_I_NODEV flag from s_iflags is both
 	 * necessary and safe.
 	 */
-	sb->s_iflags &= ~SB_I_NODEV;
+	sb->s_iflags &= ~MS_NODEV;
 	sb->s_iflags |= SB_I_NOEXEC;
 	sb->s_magic = BINDERFS_SUPER_MAGIC;
 	sb->s_op = &binderfs_super_ops;
