diff --git a/fs/kernfs/mount.c b/fs/kernfs/mount.c
index 58a9a07..1c2ea6c 100644
--- a/fs/kernfs/mount.c
+++ b/fs/kernfs/mount.c
@@ -139,9 +139,7 @@ struct dentry *kernfs_node_dentry(struct kernfs_node *kn,
 			dput(dentry);
 			return ERR_PTR(-EINVAL);
 		}
-		mutex_lock(&d_inode(dentry)->i_mutex);
 		dtmp = lookup_one_len(kntmp->name, dentry, strlen(kntmp->name));
-		mutex_unlock(&d_inode(dentry)->i_mutex);
 		dput(dentry);
 		if (IS_ERR(dtmp))
 			return dtmp;
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 4ba9512..0e9bbb2 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -5945,7 +5945,6 @@ void cgroup_exit(struct task_struct *tsk)
 	if (!list_empty(&tsk->cg_list)) {
 		spin_lock_irq(&css_set_lock);
 		css_set_move_task(tsk, cset, NULL, false);
-		spin_unlock_irq(&css_set_lock);
 		list_add_tail(&tsk->cg_list, &cset->dying_tasks);
 		cset->nr_tasks--;
 
@@ -5953,6 +5952,7 @@ void cgroup_exit(struct task_struct *tsk)
 			cgroup_freezer_frozen_exit(tsk);
 		else if (unlikely(cgroup_task_freeze(tsk)))
 			cgroup_update_frozen(task_dfl_cgroup(tsk));
+		spin_unlock_irq(&css_set_lock);
 	} else {
 		get_css_set(cset);
 	}
