--- a/drivers/kernelsu/su_mount_ns.c	2025-12-16 10:56:15.433023482 +0800
+++ b/drivers/kernelsu/su_mount_ns.c	2025-12-16 10:55:02.098996974 +0800
@@ -8,18 +8,29 @@
 #include <linux/namei.h>
 #include <linux/proc_ns.h>
 #include <linux/pid.h>
+#include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
 #include <linux/sched/task.h>
+#else
+#include <linux/sched.h>
+#endif
 #include <linux/slab.h>
 #include <linux/syscalls.h>
 #include <linux/task_work.h>
 #include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0)
+#include <linux/mount.h>
+#else
 #include <uapi/linux/mount.h>
+#endif
 
 #include "arch.h"
 #include "klog.h" // IWYU pragma: keep
 #include "ksu.h"
 #include "su_mount_ns.h"
 
+#include "kernel_compat.h"
+
 extern int path_mount(const char *dev_name, struct path *path,
                       const char *type_page, unsigned long flags,
                       void *data_page);
@@ -117,10 +128,12 @@
     fd_install(fd, ns_file);
     ret = ksu_sys_setns(fd, CLONE_NEWNS);
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5, 11, 0)
+    close_fd(fd);
+#elif LINUX_VERSION_CODE > KERNEL_VERSION(4, 19, 0)
     ksys_close(fd);
 #else
-    close_fd(fd);
+    sys_close(fd);
 #endif
 
     if (ret) {
@@ -145,7 +158,12 @@
 // individual mode , need CAP_SYS_ADMIN to perform unshare and remount
 static void ksu_mnt_ns_individual(void)
 {
-    long ret = ksys_unshare(CLONE_NEWNS);
+    long ret;
+#if LINUX_VERSION_CODE > KERNEL_VERSION(4, 19, 0)
+    ret = ksys_unshare(CLONE_NEWNS);
+#else
+    ret = sys_unshare(CLONE_NEWNS);
+#endif
     if (ret) {
         pr_warn("call ksys_unshare failed: %ld\n", ret);
         return;
