diff --git a/note7_bak/fs/proc/cmdline.c b/note7/fs/proc/cmdline.c
index bf4aafa..5006739 100644
--- a/note7_bak/fs/proc/cmdline.c
+++ b/note7/fs/proc/cmdline.c
@@ -27,8 +27,18 @@ static void proc_command_line_init(void) {
 }
 #endif
 
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+extern int susfs_spoof_cmdline_or_bootconfig(struct seq_file *m);
+#endif
+
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
+#ifdef CONFIG_KSU_SUSFS_SPOOF_CMDLINE_OR_BOOTCONFIG
+	if (!susfs_spoof_cmdline_or_bootconfig(m)) {
+		seq_putc(m, '\n');
+		return 0;
+	}
+#endif
 #ifdef CONFIG_INITRAMFS_IGNORE_SKIP_FLAG
 	seq_puts(m, proc_command_line);
 #else
